pageserver/src/page_service.rs

```diff
@@ -3300,57 +3370,342 @@ where
     }
 }
 
-/// Implements the page service over gRPC.
+/// Serves the page service over gRPC. Dispatches to PageServerHandler for request processing.
+///
+/// TODO: add trace spans, interceptors, and sampling.
+/// TODO: rename to PageServiceHandler when libpq impl is removed.
+pub struct GrpcPageServiceHandler {
+    tenant_manager: Arc<TenantManager>,
+    ctx: RequestContext,
+}
+
+impl GrpcPageServiceHandler {
+    /// Errors if the request is executed on a non-zero shard. Only shard 0 has a complete view of
+    /// relations and their sizes, as well as SLRU segments and other data.
+    #[allow(clippy::result_large_err)]
+    fn ensure_shard_zero(req: &tonic::Request<impl Any>) -> Result<(), tonic::Status> {
+        match Self::extract::<ShardIndex>(req).shard_number.0 {
+            0 => Ok(()),
+            shard => Err(tonic::Status::invalid_argument(format!(
+                "request must execute on shard zero (is shard {shard})",
+            ))),
+        }
+    }
+
+    /// Extracts the given type from the request extensions. It must have been set by an
+    /// interceptor.
+    fn extract<T: Send + Sync + 'static>(req: &tonic::Request<impl Any>) -> &T {
+        req.extensions()
+            .get::<T>()
+            .expect("extension should be set by interceptor")
+    }
+
+    /// Generates a PagestreamRequest header from a ReadLsn and request ID.
+    fn make_hdr(read_lsn: page_api::ReadLsn, req_id: u64) -> PagestreamRequest {
+        PagestreamRequest {
+            reqid: req_id,
+            request_lsn: read_lsn.request_lsn,
+            not_modified_since: read_lsn.not_modified_since_lsn.unwrap_or_default(),
```

Comment:
Why is this optional? AAIU the compute always provides some value. In case, defaulting to `Lsn(0)` doesn't seem right.

Reply:
Nice catch, thanks. Protocol-wise we don't really _need_ the client to specify it (we can just read at the request LSN), but the current behavior seems to be that it sets it equal to request LSN in that case rather than 0. I'll submit an update to do that.

https://github.com/neondatabase/neon/blob/232591e457c62de8f70b60678a85b8fc0434e5d6/pageserver/src/page_service.rs#L2232-L2241

Reply:
Addressed in https://github.com/neondatabase/neon/pull/12094/commits/637829c38b0db61ae77f16a1d7323ed6142ed7e8.